src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs

```diff
@@ -116,16 +117,31 @@ public static SocketError Connect(SafeCloseSocket handle, byte[] peerAddress, in
                 IntPtr.Zero);
             return errorCode == SocketError.SocketError ? GetLastSocketError() : SocketError.Success;
         }
-
         public static SocketError Send(SafeCloseSocket handle, IList<ArraySegment<byte>> buffers, SocketFlags socketFlags, out int bytesTransferred)
         {
+            const int StackThreshold = 16; // arbitrary limit to avoid too much space on stack (note: may be over-sized, that's OK - length passed separately)
             int count = buffers.Count;
-            WSABuffer[] WSABuffers = new WSABuffer[count];
-            GCHandle[] objectsToPin = null;
+            bool useStack = count <= StackThreshold;
+
+            WSABuffer[] leasedWSA = null;
+            GCHandle[] leasedGC = null;
+            Span<WSABuffer> WSABuffers = stackalloc WSABuffer[0];
+            Span<GCHandle> objectsToPin = stackalloc GCHandle[0];
+            if (useStack)
+            {
+                WSABuffers = stackalloc WSABuffer[StackThreshold];
+                objectsToPin = stackalloc GCHandle[StackThreshold];
+                objectsToPin.Clear(); // note touched in "finally"
+            }
+            else
+            {
+                WSABuffers = leasedWSA = ArrayPool<WSABuffer>.Shared.Rent(count);
+                objectsToPin = leasedGC = ArrayPool<GCHandle>.Shared.Rent(count);
+                Array.Clear(leasedGC, 0, count);
```

Comment:
Rather than doing both the `objectsToPin.Clear()` above and the `Array.Clear(leasedGC, 0, count)` here, after the else block you could just do:
```C#
objectsToPin.Slice(0, count).Clear();
```

Reply:
@stephentoub  I thought about that - only thing that stopped me was the potential cost of slicing; if you're happy that it has been tuned enough, then: fine